### ISSUE TITLE
reverse_dns_lookup causes havoc and destruction

### ISSUE THREAD

**Issue by danigulyas** [rabbitmq/rabbitmq-auth-backend-http#38]
Good afternoon!:)

I'm through 1.5 days of debugging and magicking, I happened to be so brave that i've enabled `reverse_dns_lookups` (set to true) at the config meanwhile using both `web-stomp` and `auth-http-backend` plugins.

Since there's no record defined for `Sock`, by enabling `reverse_dns_lookups` it was perfectly possible to pass 
```erlang
{authz_socket_info,{<<"73b85905f2f8">>,15674},{<<"172.19.0.1">>,49004}}
```
instead of
```erlang
{authz_socket_info,{{172,19,0,3},15674},{{172,19,0,1},49024}}
```
to [extract_address()](https://github.com/rabbitmq/rabbitmq-auth-backend-http/blob/master/src/rabbit_auth_backend_http.erl#L143), which would pass the bitstring (instead of the tuple) to the function `inet_parse:ntoa()`, which would throw an error since it only accepts ip addresses in a tuple format.

I have no major experience in Erlang, although i would suggest to define a record for the `Sock` so it can be verified that it's the `hostname bitstring type` or the `ip tuple type`.

*TL;DR*: People of RabbitMQ, beware, don't use `reverse_dns_lookups` with `web-stomp` and `auth-http-backend`.

Let me know if there's been a decision or so regarding what approach should be taken in order to fix it, i might be able to create a PR for it.

Cheers,
-Dan

**Comment by michaelklishin** [comment]
Records in Erlang do not perform type checks for field values (Erlang is dynamically typed) but some kind of validation can be considered. Please take this to the mailing list.

**Comment by danigulyas** [comment]
@michaelklishin Cool, thanks! In the meantime, would it be reasonable to educate users about this behaviour (by placing some kind of warning or something in `readme.md` maybe) that they shouldn't use this option when using this plugin?

**Comment by michaelklishin** [comment]
Unfortunately users do not read READMEs before using things. Again, this needs a discussion and we do not use GitHub issues for discussions. Please take this to the mailing list.

**Comment by binarin** [comment]
BTW I think it's a pretty actionable issue.

Callback spec https://github.com/rabbitmq/rabbitmq-common/blob/3885ca1967b1b340c10f7110b67fcb414f2dfda4/src/rabbit_authz_backend.erl#L49 needs to be updated to include (or be completely replaced with) `#authz_socket_info{}`.  Maybe the record itself should also be covered with specs.

And then this plugin should be taught how to handle all inputs that are covered by the specs above.
